python/lsst/dax/distribution/DataIngest.py

```diff
@@ -0,0 +1,343 @@
+            The transaction id number.
+        abort : bool
+            True if the transaction should be rolled back.
+
+        Return
+        ------
+        success : bool
+            True if the operation was successful.
+        status : int
+            Status code from the put operation.
+        r_json : json or None
+            json data from the put operation if status was 200.
+        """
+        cmd = 'ingest/trans/' + str(transaction_id) + '?abort='
+        if abort:
+            cmd += '1'
+        else:
+            cmd += '0'
+        success, status, r_json = self._putToIngest(cmd, {'auth_key':self._auth_key})
+        if not success:
+            print("ERROR ending transaction id=", transaction_id, "abort=", abort, "status=", status,
+                  "r_json=", r_json)
+        return success, status, r_json
+
+    def getChunkTargetAddr(self, transaction_id, chunk_id):
+        """ Get the host and port number of the ingest worker for this chunk.
+
+        Parameters
+        ----------
+        transaction_id : int
+            Ingest transaction id number.
+        chunk_id : int
+            Chunk id number.
+
+        Return
+        ------
+        host : str
+            Host name of the ingest worker.
+        port : int
+            Port number of the ingest worker.
+        """
+        cmd = 'ingest/chunk'
+        jdata = {"transaction_id":transaction_id,"chunk":chunk_id,"auth_key":self._auth_key}
+        success, r_json = self._postToIngest(cmd, jdata)
+        if not success:
+            print("ERROR failed to get chunk target address", jdata, " r_json=", r_json)
+            raise RuntimeError('Transaction ' + transaction_id
+                               + ' failed to get target address for chunk ' + chunk_id)
+        host = r_json['location']['host']
+        port = r_json['location']['port']
+        return host, port
+
+    def sendChunkToTarget(self, host, port, transaction_id, table, f_path):
+        """ Send the file to the ingest worker.
+
+        Parameters
+        ----------
+        host : str
+            Host name for the ingest worker.
+        port : int
+            Port number for the ingest worker.
+        transation_id : int
+            Ingest transaction id number.
+        table : str
+            The name of the table that f_path data should be added to.
+        f_path : str
+            Full path to the file to ingest.
+
+        Note
+        ----
+        This calls the external program 'qserv-replica-file-ingest' to
+        actually send the file.
+        """
+        cmd = ('qserv-replica-file-ingest FILE ' + host + ' ' + str(port) + ' '
+            + str(transaction_id) + ' ' + table + ' P ' + f_path + ' --verbose --columns-separator=TAB')
+        print("cmd=", cmd)
+        process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        out_str = process.communicate()
+        process.wait()
```

Comment:
Why we need to use subprocess() for a synchronous call?

Reply:
In modern python you should use `subprocess.run()` directly. Much simpler API.

Reply:
Also note you'll get bytes back so will need to decode to str.

Reply:
I'm finding trying to use subprocess.run really annoying. The cmd argument that works for subprocess.Popen doesn't work with subprocess.run. I've tried a few permutations and they've all failed, which I think is why I gave up on run in the first place.
`process = subprocess.run(cmd, cwd=cwd, shell=True, universal_newlines=True,
                                 stdout=subprocess.PIPE, stderr=subprocess.STDOUT)`
`process = subprocess.run([cmd, args], cwd=cwd, shell=True, universal_newlines=True,
                                 stdout=subprocess.PIPE, stderr=subprocess.STDOUT)`

I'm trying to execute commands like 
`python ~/work/dax_data_generator/bin/datagen.py  --chunk 1201 --visits 30 --objects 1000 gencfg.py`
    subprocess.run consistently complains that it can't find datagen.py
Or 
` sph-partition -c /home/jgates/work/dax_data_generator/python/lsst/dax/distribution/fakeData/partitionCfgs/Object.cfg --mr.num-workers 1 --out.dir outdirObject --in chunk0_CT_Object.csv --in chunk402_CT_Object.csv --in chunk401_CT_Object.csv --in chunk400_CT_Object.csv --in chunk404_EO_Object.csv --in chunk403_CT_Object.csv`
Reconstructing them as lists is a nuisance.

Having the whole command line in one string makes it really easy to debug failures as I just cut and paste the string in a terminal.

Reply:
You may well find that `~` is not expanded by the subprocess command depending on the shell setting. It should be able to take a single string because it is documented to take the normal Popen commands but with a much simplified interface.

Having all the commands separately is much more secure and automatically deals with any problems with shell quoting.

Reply:
I did the same thing I did yesterday and today it worked. I'm at a loss.